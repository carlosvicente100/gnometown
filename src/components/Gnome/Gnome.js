import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Link } from '@reach/router'

const CardContainer = styled.div`
  font-family: Helvetica;
  background-color: #edeff6;
  color: black;
  border-radius: 25px;
  overflow: hidden;
  img {
    border: 1px solid black;
    grid-area: image;
  }
  ul,
  h3 {
    margin: 2px;
    text-align: left;
  }
  .name {
    grid-area: gnome-name;
    margin: auto;
  }
  .properties {
    display: block;
    grid-area: properties;
    font-size: 14px;
    text-align: left;
    span {
      font-weight: bold;
    }
    p {
      margin: auto;
    }
  }
  .professions {
    display: block;
    grid-area: jobs;
    font-size: 14px;
  }
  .friends {
    display: block;
    grid-area: friends;
    font-size: 14px;
  }
  display: grid;
  height: 400px;
  grid-template-areas:
  '. image .'
  '. gnome-name .'
  '. properties .'
  '. jobs .'
  '. friends .';
  width: 230px;
  margin: auto;
  //margin: 20px;
  


  @media (max-width: 1201px) {
    font-size: 13px;
    width: 100%;
    height: 80px;
    grid-template-columns: auto 80px auto auto auto;
    grid-template-rows: auto;
    .name{
        font-size: 10px;

      }
    }
    .properties,
    .professions,
    .friends {
      display: none;
    }

    .properties-small,
    .professions-small,
    .friends-small {
      margin: auto;
      display: inline-flex;
      i{
        margin: 0px 10px;
      }
    }
    grid-template-areas: 'image gnome-name properties jobs friends';
    display: inline-grid;
    img {
      width: 50px;
      height: 50px;
    }
  }
  @media (min-width: 1201px) {
    .properties-small,
    .professions-small,
    .friends-small {
      margin: auto;
      display: none;
    }

    img {
      width: 90px;
      height: 90px;
    }

    .name {
      font-size: 14px;
    }
    .properties {
      display: block;
    }
    .professions {
      display: block;
    }
    .friends {
      display: block;
    }
    grid-template-areas:
      '. image .'
      '. gnome-name .'
      '. properties .'
      '. jobs .'
      '. friends .';
  }
`

const CardImage = styled.img`
  height: 90px;
  margin: auto;
  width: 90px;
  border-radius: 100px;
`

const Card = ({ name, age, thumbnail, height, weight, friends, professions, id }) => {
  const ReduceList = (list, maxElements = 3) => {
    const newList = list.map((element, index) => {
      if (index < maxElements) {
        return <li key={index}>{element}</li>
      } else if (index === maxElements) {
        return <li key={index}>and more...</li>
      }
    })
    return newList
  }

  return (
    <Link to={`/profile/${id}`}>
      <CardContainer>
        <CardImage src={thumbnail}></CardImage>
        <h2 className="name">{name}</h2>
        <div className="properties">
          <p>
            <span>Age:</span>
            {` ${age} years`}
          </p>
          <p>
            <span>Height:</span>
            {`${parseFloat(height).toFixed(2)} cm`}
          </p>
          <p>
            <span>Weight:</span>
            {` ${parseFloat(weight).toFixed(2)} kg`}
          </p>
        </div>
        <div className="properties-small">
          <div className="age">
            <i class="fas fa-hourglass-end"></i>
            {` ${age} `}
          </div>
          <div className="height">
            <i class="fas fa-arrows-alt-v"></i>
            {`${parseFloat(height).toFixed(2)}`}
          </div>
          <div className="weight">
            <i class="fas fa-arrows-alt-h"></i>
            {` ${parseFloat(weight).toFixed(2)} `}
          </div>
        </div>
        <div className="professions">
          <h3>Professions</h3>
          {professions.length > 0 ? <ul>{ReduceList(professions)}</ul> : <span>¯\_(ツ)_/¯</span>}
        </div>
        <div className="professions-small">
          <i class="fas fa-hammer"></i>
          <span>{professions.length > 0 ? professions.length : 0} </span>
        </div>
        <div className="friends">
          <h3>Friends</h3>
          {friends.length > 0 ? <ul>{ReduceList(friends)}</ul> : <span>¯\_(ツ)_/¯</span>}
        </div>
        <div className="friends-small">
          <i class="fas fa-users"></i>
          <span>{friends.length > 0 ? friends.length : 0} </span>
        </div>
      </CardContainer>
    </Link>
  )
}
Card.propTypes = {
  name: PropTypes.string.isRequired,
  age: PropTypes.number.isRequired,
  thumbnail: PropTypes.string.isRequired,
  height: PropTypes.number.isRequired,
  weight: PropTypes.number.isRequired,
  friends: PropTypes.array,
  professions: PropTypes.array
}

Card.defaultProps = {}

export default Card
