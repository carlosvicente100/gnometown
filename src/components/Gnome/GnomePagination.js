import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

const GnomePaginationContainer = styled.button`
  margin: 0;
  border: 0;
  width: 100%;
  background-color: #f8f8f8;
  height: 60px;
  cursor: pointer;
  padding: 10px 30px 10px 10px;
  display: flex;
  justify-content: center;
  border-bottom: 1px solid #e3e3e3;
  align-items: center;
  &:hover {
    background-color: #f6e58d;
  }
  label {
    font-family: Helvetica;
    color: #707070;
    font-size: 16px;
  }
`

const GnomePagination = ({ text, onClick }) => {
  const onClickContainer = () => {
    onClick()
  }
  return (
    <GnomePaginationContainer onClick={onClickContainer}>
      <label data-testid="more-gnomes">{text}</label>
    </GnomePaginationContainer>
  )
}

GnomePagination.propTypes = {
  text: PropTypes.string,
  onClick: PropTypes.func
}

GnomePagination.defaultProps = {
  text: 'Load more gnomes',
  onClick: () => {}
}

export default GnomePagination
