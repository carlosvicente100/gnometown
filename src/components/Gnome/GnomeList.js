import React, { useState } from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

import Gnome from './Gnome'
import GnomePagination from './GnomePagination'

const ItemListContainer = styled.div`
  display: grid;
  grid-template-columns: auto;
  margin: auto;
  column-gap: 20px;
  row-gap: 5px;
  margin: 20px;

  a {
    text-decoration: none;
    width: fit-content;
    margin: auto;

    margin: 0;
    width: 100%;
  }

  @media (min-width: 688px) {
  }
  @media (min-width: 900px) {
  }
  @media (min-width: 1201px) {
    row-gap: 20px;
    grid-template-columns: auto auto auto auto;
    a {
      width: fit-content;
      margin: auto;
    }
  }
  @media (min-width: 1501px) {
    grid-template-columns: auto auto auto auto auto;
  }
`

const GnomeList = ({ gnomes, gnomesPerPage = 5 }) => {
  const [page, setPage] = useState(1)

  const loadMoreUsers = () => {
    setPage(page + 1)
  }

  const moreUsers = gnomesPerPage * page >= gnomes.length ? false : true

  console.log('/// gnomes', gnomes)
  console.log('/// gnomesPerPage', gnomesPerPage)

  return (
    <>
      <ItemListContainer className="GnomeList">
        {gnomes.length > 0 ? (
          gnomes.filter((a, index) => index < gnomesPerPage * page).map((gnome) => <Gnome key={gnome.id} {...gnome} />)
        ) : (
          //  gnomes.map((gnome) => <Gnome key={gnome.id} {...gnome} />)
          <div>no gnomes founded :(</div>
        )}
      </ItemListContainer>
      {moreUsers && <GnomePagination onClick={loadMoreUsers} />}
    </>
  )
}

GnomeList.propTypes = {
  gnomes: PropTypes.array.isRequired
}
GnomeList.defaultProps = {}

export default GnomeList
