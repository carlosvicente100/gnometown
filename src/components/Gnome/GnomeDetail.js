import React from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'

import GnomeList from '../Gnome/GnomeList'

const CardContainer = styled.div`
  background-color: #7f8fa6;
  font-family: 'Helvetica';
  color: black;
  border-radius: 25px;
  overflow: hidden;
  height: 100%;
  display: grid;
  column-gap: 10px;
  img {
    grid-area: image;
  }
  ul,
  h3 {
    margin: 2px;
  }
  h2 {
    height: auto;
    grid-area: title;
    font-size: 16px;
    text-transform: uppercase;
  }
  .gnome-professions {
    margin: 20px 0px;
    grid-area: jobs;
    font-size: 14px;
    ul {
      list-style: none;
      text-align: left;
      width: 20%;
      margin: auto;
      text-align: left;
      li {
        width: fit-content;
      }
    }
  }
  .gnome-friends {
    grid-area: friends;
    text-align: center;
    font-size: 14px;
    .GnomeList {
      margin: 0px;
    }
  }
  text-align: center;
  grid-row-gap: 20px;
  grid-template-areas:
    '.image .'
    '. title .'
    '. properties .'
    '. jobs .'
    'friends friends friends';

  .properties {
    grid-area: properties;
    font-size: 14px;
    p {
      display: flex;
      justify-content: space-evenly;
    }
    span {
      font-weight: bold;
    }
  }
  padding: 20px;
  margin: auto;
`

const CardImage = styled.img`
  height: 200px;
  margin: auto;
  width: 200px;
  border-radius: 100px;
`

const HairContainer = styled.div`
  display: inline-flex;
  margin: auto;
  width: fit-content;
  .dummy {
    margin: auto;
    margin-left: 5px;
    width: 20px;
    height: 20px;
    border-radius: 50%;
    border: 2px solid black;
  }
`
const GnomeDetail = ({ name, age, thumbnail, height, weight, friends, professions, hair_color }) => {
  return (
    <CardContainer>
      <CardImage src={thumbnail} data-testid="detail-image"></CardImage>
      <h2>{name}</h2>
      <div className="properties" data-testid="detail-properties">
        <p>
          <span>Age: </span>
          {`${age} years`}
        </p>
        <p>
          <span>Height:</span>
          {`${parseFloat(height).toFixed(2)} cm`}
        </p>
        <p>
          <span>Weight:</span>
          {` ${parseFloat(weight).toFixed(2)} kg`}
        </p>
        <HairContainer>
          <p>Hair Color: </p>
          <div className="dummy" style={{ backgroundColor: hair_color }} />
        </HairContainer>
      </div>

      {professions.length > 0 && (
        <div className="gnome-professions" data-testid="detail-professions">
          <h3>Professions</h3>
          <ul>
            {professions.map((element, index) => (
              <li key={index}>{element}</li>
            ))}
          </ul>
        </div>
      )}
      {friends.length > 0 && (
        <div className="gnome-friends" data-testid="detail-friends">
          <h3>Friends</h3>
          <GnomeList gnomes={friends}></GnomeList>
        </div>
      )}
    </CardContainer>
  )
}
GnomeDetail.propTypes = {
  name: PropTypes.string.isRequired,
  age: PropTypes.number.isRequired,
  thumbnail: PropTypes.string.isRequired,
  hair_color: PropTypes.string.isRequired,
  height: PropTypes.number.isRequired,
  weight: PropTypes.number.isRequired,
  friends: PropTypes.array,
  professions: PropTypes.array
}

GnomeDetail.defaultProps = {}

export default GnomeDetail
