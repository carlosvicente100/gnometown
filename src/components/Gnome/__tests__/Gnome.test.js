import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Gnome from '../Gnome'

const dummyProps = {
  id: 1337,
  name: 'dummy Name',
  thumbnail: 'http://placekitten.com/200/300',
  age: 240,
  weight: 40.97596,
  height: 127.88554,
  hair_color: 'Red',
  professions: ['Carpenter', 'Farmer', 'Stonecarver', 'Brewer', 'Tax inspector', 'Prospector'],
  friends: ['Sarabink Tinkbuster', 'Tinadette Wrongslicer']
}

describe('Gnome Component Suite', () => {
  test('Snapshot check', () => {
    const component = renderer.create(<Gnome {...dummyProps} />)
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Should have Link - prop id', () => {
    const { getByTestId } = render(<Gnome {...dummyProps} />)
    const gnomeLink = getByTestId('gnome-url')
    expect(gnomeLink.getAttribute('href')).toBe(`/profile/${dummyProps.id}`)
  })

  test('Should use simply props correcty - name, age, height, weight', () => {
    const { getByText } = render(<Gnome {...dummyProps} />)
    expect(getByText(dummyProps.name)).toBeInTheDocument()
    expect(getByText(`${dummyProps.age} years`)).toBeInTheDocument()
    expect(getByText(`${parseFloat(dummyProps.height).toFixed(2)} cm`)).toBeInTheDocument()
    expect(getByText(`${parseFloat(dummyProps.weight).toFixed(2)} kg`)).toBeInTheDocument()
  })
  test('Should use image prop correcty', () => {
    const { getByTestId } = render(<Gnome {...dummyProps} />)
    const gnomeImage = getByTestId('gnome-image')
    expect(gnomeImage.getAttribute('src')).toBe(dummyProps.thumbnail)
  })

  test('Should list professions prop correcty, only 3, limited size', () => {
    const { getByTestId } = render(<Gnome {...dummyProps} />)
    const professions = getByTestId('gnome-professions')
    expect(professions.querySelector('h3').innerHTML).toBe('Professions')

    const professionList = professions.children[1]

    expect(professionList.children.length).toBe(4)
    expect(professionList.children[0].className === 'job-element')
    expect(professionList.children[0].innerHTML).toBe(dummyProps.professions[0])
    expect(professionList.children[1].innerHTML).toBe(dummyProps.professions[1])
    expect(professionList.children[2].innerHTML).toBe(dummyProps.professions[2])
    expect(professionList.children[3].innerHTML).toBe('and more...')
  })

  test('Should list friends prop correcty', () => {
    const { getByTestId } = render(<Gnome {...dummyProps} />)
    const friends = getByTestId('gnome-friends')
    expect(friends.querySelector('h3').innerHTML).toBe('Friends')

    const friendList = friends.children[1]

    expect(friendList.children.length).toBe(2)
    expect(friendList.children[0].className === 'friend-element')
    expect(friendList.children[0].innerHTML).toBe(dummyProps.friends[0])
    expect(friendList.children[1].innerHTML).toBe(dummyProps.friends[1])
  })
})

describe('Gnome Component Suite - exceptions', () => {
  test('Should render without friends ', () => {
    const { getByTestId } = render(<Gnome {...dummyProps} friends={[]} />)
    const friends = getByTestId('gnome-friends')
    expect(friends.querySelector('span').innerHTML).toBe('¯\\_(ツ)_/¯')
  })

  test('Should render without professions ', () => {
    const { getByTestId } = render(<Gnome {...dummyProps} professions={[]} />)
    const friends = getByTestId('gnome-professions')
    expect(friends.querySelector('span').innerHTML).toBe('¯\\_(ツ)_/¯')
  })
})
