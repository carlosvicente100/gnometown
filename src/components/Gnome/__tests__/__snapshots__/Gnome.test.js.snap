// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gnome Component Suite Snapshot check 1`] = `
.c0 {
  font-family: Helvetica;
  background-color: #edeff6;
  color: black;
  border-radius: 25px;
  overflow: hidden;
  display: grid;
  height: 400px;
  grid-template-areas: '. image .' '. gnome-name .' '. properties .' '. jobs .' '. friends .';
  width: 230px;
  margin: auto;
  grid-template-areas: 'image gnome-name properties jobs friends';
  display: inline-grid;
}

.c0 img {
  border: 1px solid black;
  grid-area: image;
}

.c0 ul,
.c0 h3 {
  margin: 2px;
  text-align: left;
}

.c0 .name {
  grid-area: gnome-name;
  text-align: center;
  margin: auto;
}

.c0 .properties {
  display: block;
  grid-area: properties;
  font-size: 14px;
  text-align: left;
}

.c0 .properties span {
  font-weight: bold;
}

.c0 .properties p {
  margin: auto;
}

.c0 .professions {
  display: block;
  grid-area: jobs;
  font-size: 14px;
}

.c0 .friends {
  display: block;
  grid-area: friends;
  font-size: 14px;
}

.c0 .properties,
.c0 .professions,
.c0 .friends {
  display: none;
}

.c0 .properties-small {
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  text-align: center;
}

.c0 .properties-small,
.c0 .professions-small,
.c0 .friends-small {
  margin: auto 5px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c0 .properties-small i,
.c0 .professions-small i,
.c0 .friends-small i {
  margin: 0px 10px;
}

.c0 img {
  width: 40px;
  height: 40px;
  margin-left: 5px;
}

.c1 {
  height: 90px;
  margin: auto;
  width: 90px;
  border-radius: 100px;
}

@media (max-width:1201px) {
  .c0 {
    font-size: 13px;
    width: 100%;
    height: 80px;
    grid-template-columns: auto 80px auto auto auto;
    grid-template-rows: auto;
  }

  .c0 .name {
    font-size: 10px;
    margin-left: 5px;
    text-align: justify;
  }
}

@media (min-width:1201px) {
  .c0 {
    grid-template-areas: '. image .' '. gnome-name .' '. properties .' '. jobs .' '. friends .';
  }

  .c0 .properties-small,
  .c0 .professions-small,
  .c0 .friends-small {
    margin: auto;
    display: none;
  }

  .c0 img {
    margin: auto;
    width: 90px;
    height: 90px;
  }

  .c0 .name {
    font-size: 14px;
  }

  .c0 .professions,
  .c0 .friends,
  .c0 .properties {
    display: block;
  }

  .c0 .professions ul,
  .c0 .friends ul,
  .c0 .properties ul {
    list-style: none;
  }
}

<a
  data-testid="gnome-url"
  href="/profile/1337"
  onClick={[Function]}
>
  <div
    className="c0"
  >
    <img
      className="c1"
      data-testid="gnome-image"
      src="http://placekitten.com/200/300"
    />
    <h2
      className="name"
      data-testid="gnome-name"
    >
      dummy Name
    </h2>
    <div
      className="properties"
    >
      <p
        data-testid="gnome-age"
      >
        <span>
          Age:
        </span>
         240 years
      </p>
      <p
        data-testid="gnome-height"
      >
        <span>
          Height:
        </span>
        127.89 cm
      </p>
      <p
        data-testid="gnome-weight"
      >
        <span>
          Weight:
        </span>
         40.98 kg
      </p>
    </div>
    <div
      className="properties-small"
    >
      <div
        className="age"
        data-testid="gnome-small-age"
      >
        <i
          className="fas fa-birthday-cake"
        />
        <span>
           240 
        </span>
      </div>
      <div
        className="height"
        data-testid="gnome-height"
      >
        <i
          className="fas fa-arrows-alt-v"
        />
        <span>
          127.89
        </span>
      </div>
      <div
        className="weight"
        data-testid="gnome-small-weight"
      >
        <i
          className="fas fa-weight-hanging"
        />
        <span>
           40.98 
        </span>
      </div>
    </div>
    <div
      className="professions"
      data-testid="gnome-professions"
    >
      <h3>
        Professions
      </h3>
      <ul>
        <li
          className="job-element"
        >
          Carpenter
        </li>
        <li
          className="job-element"
        >
          Farmer
        </li>
        <li
          className="job-element"
        >
          Stonecarver
        </li>
        <li
          className="job-element"
        >
          and more...
        </li>
      </ul>
    </div>
    <div
      className="professions-small"
      data-testid="gnome-small-professions"
    >
      <i
        className="fas fa-hammer"
      />
      <span>
        6
         
      </span>
    </div>
    <div
      className="friends"
      data-testid="gnome-friends"
    >
      <h3>
        Friends
      </h3>
      <ul>
        <li
          className="friend-element"
        >
          Sarabink Tinkbuster
        </li>
        <li
          className="friend-element"
        >
          Tinadette Wrongslicer
        </li>
      </ul>
    </div>
    <div
      className="friends-small"
      data-testid="gnome-small-friends"
    >
      <i
        className="fas fa-users"
      />
      <span>
        2
         
      </span>
    </div>
  </div>
</a>
`;
